src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -113,22 +113,63 @@ It is implemented recursively to ensure type stability.
     return self + row + _interactions(as, rest)
 end
 
+"""
+    trap_potential(add::SingleComponentFockAddress, pot)
+
+Calculate trap potential energy of a single-component address `add`. The (precomputed) 
+potential energy per particle at each point in the lattice is given by the vector `pot`.
+```math
+V = \\sum_{d=1}^D \\sum_{i=1}^M V_d x_{i,d}^2 n_i,
+```
+where ``x = (x_1, ..., x_D)`` is the distance of site `i` to the centre of the trap and 
+`n_i` is the number of particles at site `i`. `M` is the total number of sites on the 
+lattice.
+"""
+function trap_potential(add::SingleComponentFockAddress, pot)
+    pe = 0.0
+    for (n,i) in occupied_modes(add)
+        pe += n * pot[i]
+    end
+    return pe
+end
+
+"""
+    trap_potential(add::CompositeFS, pot::Matrix)
+
+Calculate potential energy of a multicomponent address as the sum of the potential energy
+of each component. Each column of `pot` is a precomputed vector of potential energy for 
+that component at each lattice point.
+"""
+function trap_potential(add::CompositeFS, pot::Matrix)
+    pe = 0.0
+    for (i,c) in enumerate(add.components)
+        pe += trap_potential(c, @view pot[:,i])
+    end
+    return pe
+end
+
 ###
 ### HubbardRealSpace
 ###
 """
-    HubbardRealSpace(address; u=ones(C, C), t=ones(C), geometry=PeriodicBoundaries(M,))
+    HubbardRealSpace(address; geometry=PeriodicBoundaries(M,), t=ones(C), u=ones(C, C), v=zeros(C, D))
 
 Hubbard model in real space. Supports single or multi-component Fock state
 addresses (with `C` components) and various (rectangular) lattice geometries
-in arbitrary dimensions.
+in `D` dimensions.
 
 ```math
   \\hat{H} = -\\sum_{\\langle i,j\\rangle,σ} t_σ a^†_{iσ} a_{jσ} +
   \\frac{1}{2}\\sum_{i,σ} u_{σ,σ} n_{iσ} (n_{iσ} - 1) +
   \\frac{1}{2}\\sum_{i,σ≠τ}u_{σ,τ} n_{iσ} n_{iτ}
 ```
 
+If `v` is nonzero then this calculates ``\\hat{H}+\\hat{V}`` where
+```math
+    \\hat{V} = \\sum_{σ=1}^C \\sum_{d=1}^D \\sum_{i=1}^M v_{σ,d} x_{σ,i,d}^2 n_{iσ}
+```
+is the external harmonic trapping potential.
```

Comment:
Hmm, this looks a little strange, as there is a new index `d` in this line for counting the dimension that does not appear in the top line. If you use it, it will be needed to fully index the number operator as well. It is unclear what `x_{σ,i,d}` stands for. Also, grids don't have to be square (i.e. use the same number of modes).

In the first line `i` stands for any lattice sites and `\\langle i,j\\rangle` stands for nearest neighbours. Consistent with that would be to sum over all lattice sites `i` and make `x` a function that figures out what the distance from the centre is in the `x` direction. Maybe something like this:

```suggestion
If `v` is nonzero then this calculates ``\\hat{H}+\\hat{V}`` where
#```math
   \\hat{V} = \\sum_{i,σ}  (v_{σx} x(i)^2 + [v_{σy} y(i)^2 + v_{σz} z(i)^2]) n_{iσ} ,
#```
is the harmonic external potential, which depends quadratically 
on the distances from the central lattice point (in number of lattice sites) ``x(i)``, ``y(i)``, or ``z(i)`` in the 
respective dimension.
```

Unfortunately, GitHub got confused with the highlighting here.